proof/hash_to_g1.saw

```diff
@@ -0,0 +1,163 @@
+import "../spec/implementation/HashToG1.cry";
+EC <- cryptol_load "../spec/ShortWeierstrassCurve.cry";
+
+// Specs
+
+let blst_hash_to_g1_spec msg_len DST_len aug_len = do {
+  p_ptr <- llvm_alloc POINTonE1_type;
+  (msg, msg_ptr) <- ptr_to_fresh_readonly "msg" (llvm_array msg_len (llvm_int 8));
+  (DST, DST_ptr) <- ptr_to_fresh_readonly "DST" (llvm_array DST_len (llvm_int 8));
+  (aug, aug_ptr) <- ptr_to_fresh_readonly "aug" (llvm_array aug_len (llvm_int 8));
+  llvm_execute_func [p_ptr, msg_ptr, llvm_term {{ `msg_len : [64] }}, DST_ptr, llvm_term {{ `DST_len : [64] }}, aug_ptr, llvm_term {{ `aug_len : [64] }}];
+  llvm_points_to p_ptr (llvm_term {{ POINTonE1_rep (hash_to_curve_opt (aug#msg, DST)) }});
+};
+
+let map_to_g1_nonnull_spec = do {
+  out_ptr <- llvm_alloc POINTonE1_type;
+  (u,u_ptr) <- ptr_to_fresh_readonly "u" vec384_type;
+  (v,v_ptr) <- ptr_to_fresh_readonly "v" vec384_type;
+  llvm_execute_func [out_ptr, u_ptr, v_ptr];
+  llvm_points_to out_ptr (llvm_term {{ POINTonE1_rep (map_to_g1 [fp_abs u, fp_abs v]) }});
+};
+
+let map_to_isogenous_E1_spec = do {
+  p_ptr <- llvm_alloc POINTonE1_type;
+  (u, u_ptr) <- ptr_to_fresh_readonly "u" vec384_type;
+  llvm_execute_func [p_ptr, u_ptr];
+  llvm_points_to p_ptr (llvm_term {{ POINTonE1_rep (map_to_curve_simple_swu (fp_abs u)) }});
+};
+
+let isogeny_map_to_E1_alias_spec = do {
+  (out, out_ptr) <- ptr_to_fresh "out" POINTonE1_type;
+  llvm_execute_func [out_ptr, out_ptr];
+  llvm_points_to out_ptr (llvm_term {{ POINTonE1_rep (iso_map (POINTonE1_abs out)) }});
+};
+
+let dbl_n_add_spec n = do {
+  (out, out_ptr) <- ptr_to_fresh "out" POINTonE1_type;
+  (p, p_ptr) <- ptr_to_fresh_readonly "p" POINTonE1_type;
+  llvm_precond {{ POINTonE1_invariant out /\ POINTonE1_invariant p }};
+  llvm_execute_func [out_ptr, llvm_term {{`n:[64]}}, p_ptr];
+  llvm_points_to out_ptr (llvm_term {{ POINTonE1_rep (dbl_n_add (POINTonE1_abs out) `n (POINTonE1_abs p)) }});
+};
+
+// Proofs
+
+map_to_isogenous_E1_ov <- admit "map_to_isogenous_E1" map_to_isogenous_E1_spec;
+isogeny_map_to_E1_alias_ov <- admit "isogeny_map_to_E1" isogeny_map_to_E1_alias_spec;
+
+enable_experimental;
+
+let verify_dbl_n_add n = custom_verify "POINTonE1_dbl_n_add" curve_operations_e1_ovs (dbl_n_add_spec n)
+  (do {
+    simplify (addsimps POINTonE1_thms (cryptol_ss ()));
+    goal_num_ite 3
+      (do {
+        unfolding ["dbl_n_add"];
+        simplify remove_higher_order_function_simpset;
+        w4_unint_z3 (foldr concat [fp_unints, ec_mult_unints] ["point_double_fp","point_add_fp","POINTonE1_rep","Fp"]);
+      })
+      z3;
+  });
+
+//verify_dbl_n_add 2;
+let dbl_n_add_ns = [2,3,9,32,16];
+dbl_n_add_ovs <- for dbl_n_add_ns verify_dbl_n_add;
+
+let a4_rep = parse_core "[bvNat 64 3415322872136444497, bvNat 64 9675504606121301699, bvNat 64 13284745414851768802, bvNat 64 2873609449387478652, bvNat 64 2897906769629812789, bvNat 64 1536947672689614213]"; // this is the representation of a4
```

Comment:
These constants can be written in Cryptol, if you rewrite with `cryptol_ss()`.  There are examples in `curve_operations.saw` and other places.  i.e. `prove_cryptol (rewrite (cryptol_ss()) {{ fp_invariant [1,2,3,4] == True }})`.